File: bench/scala/src/main/scala/ming/Environment.scala
package ming

import scala.collection.mutable

final private[ming] class BindingCell(var value: Value)

final private[ming] class Environment private (
  val parent: Option[Environment],
  initialBindings: Iterable[(String, BindingCell)],
  initialMacros: Iterable[(String, MacroDefinition)]
):
  private val bindings = mutable.LinkedHashMap.from(initialBindings)
  private val macros   = mutable.LinkedHashMap.from(initialMacros)

  def define(name: String, value: Value): Unit =
    bindings.get(name) match
      case Some(cell) =>
        cell.value = value
      case None =>
        bindings.update(name, BindingCell(value))

  def reserve(name: String): Unit =
    bindings.get(name) match
      case Some(cell) =>
        cell.value = UninitializedValue
      case None =>
        bindings.update(name, BindingCell(UninitializedValue))

  def defineAlias(name: String, cell: BindingCell): Unit =
    bindings.update(name, cell)

  def defineMacro(name: String, macroDefinition: MacroDefinition): Unit =
    macros.update(name, macroDefinition)

  def assign(name: String, value: Value, position: Position): Unit =
    lookupCell(name, position).value = value

  def lookup(name: String, position: Position): Value =
    lookupCell(name, position).value match
      case UninitializedValue =>
        SchemeFailure.raise(s"uninitialized symbol: $name", position)
      case value =>
        value

  def lookupMacro(name: String): Option[MacroDefinition] =
    macros.get(name) match
      case some @ Some(_) =>
        some
      case None =>
        parent match
          case Some(outer) => outer.lookupMacro(name)
          case None        => None

  def lookupCellOption(name: String): Option[BindingCell] =
    bindings.get(name) match
      case some @ Some(_) =>
        some
      case None =>
        parent.flatMap(_.lookupCellOption(name))

  private def lookupCell(name: String, position: Position): BindingCell =
    lookupCellOption(name) match
      case Some(cell) => cell
      case None       => SchemeFailure.raise(s"unbound symbol: $name", position)

private[ming] object Environment:

  def root(bindings: Iterable[(String, Value)] = Nil): Environment =
    new Environment(None, bindings.map((name, value) => name -> BindingCell(value)), Nil)

  def child(parent: Environment, bindings: Iterable[(String, Value)] = Nil): Environment =
    new Environment(
      Some(parent),
      bindings.map((name, value) => name -> BindingCell(value)),
      Nil
    )

File: bench/java/src/main/java/ming/Environment.java
package ming;

import java.util.HashMap;
import java.util.Map;

final class Environment {
    private final Environment parent;
    private final Map<String, BindingCell> bindings;
    private final Map<String, SyntaxRulesMacro> macros;

    Environment(Environment parent) {
        this.parent = parent;
        this.bindings = new HashMap<>();
        this.macros = new HashMap<>();
    }

    void define(String name, Value value) {
        defineCell(name, new BindingCell(value));
    }

    void defineCell(String name, BindingCell cell) {
        bindings.put(name, cell);
    }

    void defineMacro(String name, SyntaxRulesMacro definition) {
        macros.put(name, definition);
    }

    Value lookup(String name, SourceLoc loc) throws EvalError {
        BindingCell cell = lookupCell(name);
        if (cell != null) {
            return cell.value();
        }
        throw SchemeErrors.at(loc, "unbound variable: " + name);
    }

    void set(String name, Value value, SourceLoc loc) throws EvalError {
        BindingCell cell = lookupCell(name);
        if (cell != null) {
            cell.set(value);
            return;
        }
        throw SchemeErrors.at(loc, "unbound variable: " + name);
    }

    BindingCell lookupCell(String name) {
        BindingCell cell = bindings.get(name);
        if (cell != null) {
            return cell;
        }
        if (parent != null) {
            return parent.lookupCell(name);
        }
        return null;
    }

    SyntaxRulesMacro lookupMacro(String name) {
        SyntaxRulesMacro definition = macros.get(name);
        if (definition != null) {
            return definition;
        }
        if (parent != null) {
            return parent.lookupMacro(name);
        }
        return null;
    }
}

final class BindingCell {
    private Value value;

    BindingCell(Value value) {
        this.value = value;
    }

    Value value() {
        return value;
    }

    void set(Value value) {
        this.value = value;
    }
}
